plugins/teststeps/awsfileupload/awsFileUpload.go

```diff
@@ -0,0 +1,299 @@
+	Msg string
+}
+
+// Name returns the plugin name.
+func (ts FileUpload) Name() string {
+	return Name
+}
+
+func emitEvent(ctx xcontext.Context, name event.Name, payload interface{}, tgt *target.Target, ev testevent.Emitter) error {
+	payloadStr, err := json.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("cannot encode payload for event '%s': %v", name, err)
+	}
+	rm := json.RawMessage(payloadStr)
+	evData := testevent.Data{
+		EventName: name,
+		Target:    tgt,
+		Payload:   &rm,
+	}
+	if err := ev.Emit(ctx, evData); err != nil {
+		return fmt.Errorf("cannot emit event EventURLtoReport: %v", err)
+	}
+	return nil
+}
+
+// Run executes the awsFileUpload.
+func (ts *FileUpload) Run(ctx xcontext.Context, ch test.TestStepChannels, params test.TestStepParameters,
+	ev testevent.Emitter, resumeState json.RawMessage) (json.RawMessage, error) {
+	// Validate the parameter
+	if err := ts.validateAndPopulate(params); err != nil {
+		return nil, err
+	}
+	f := func(ctx xcontext.Context, target *target.Target) error {
+		// expand args
+		path, err := ts.path.Expand(target)
+		if err != nil {
+			return fmt.Errorf("failed to expand argument '%s': %v", ts.path, err)
+		}
+		filename, err := ts.filename.Expand(target)
+		if err != nil {
+			return fmt.Errorf("failed to expand argument dir '%s': %v", ts.filename, err)
+		}
+		var url string
+		// Compress if compress parameter is true
+		if ts.compress {
+			// Create buffer for the compressed data
+			var buf bytes.Buffer
+			// Create the archive and write the output to the "out" Writer
+			buf, err = createTarArchive(path, buf)
+			if err != nil {
+				return fmt.Errorf("error creating an archive: %v", err)
+			}
+			fmt.Println("Tar archive created successfully")
+			filename = filename + ".tar.gz"
+			// Upload file
+			url, err = ts.upload(filename, buf.Bytes())
+			if err != nil {
+				return fmt.Errorf("could not upload the file: %v", err)
+			}
+		} else {
+			// Read the file that should be uploaded
+			bodyBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return fmt.Errorf("could not read the file: %v", err)
+			}
+			// Upload file
+			url, err = ts.upload(filename, bodyBytes)
+			if err != nil {
+				return fmt.Errorf("could not upload the file: %v", err)
+			}
+		}
+		// Emit URLtoReport event to get the url into the report
+		if ts.emitURLtoReport {
+			if err := emitEvent(ctx, EventURLtoReport, eventURLtoReportPayload{Msg: url}, target, ev); err != nil {
+				return fmt.Errorf("failed to emit event: %v", err)
+			}
+		}
+		return nil
+	}
+	return teststeps.ForEachTarget(Name, ctx, ch, f)
+}
+
+// Retrieve all the parameters defines through the jobDesc
+func (ts *FileUpload) validateAndPopulate(params test.TestStepParameters) error {
+	// Retrieving parameter as json Raw.Message
+	// validate path and filename
+	ts.path = params.GetOne("path")
+	if ts.path.IsEmpty() {
+		return fmt.Errorf("invalid or missing 'path' parameter, must be exactly one string")
```

Comment:
a better error message, here and below, could be "missing or empty 'path' parameter". `IsEmpty` returns `true` if the parameter is not specified, or if it's an empty string, and "must be exactly one string" could be misleading